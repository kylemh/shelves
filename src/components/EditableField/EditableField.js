import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Button from '../Button/Button';
import ScreenReaderOnly from '../ScreenReaderOnly/ScreenReaderOnly';
import { ReactComponent as IconPencil } from '../../images/FontAwesomeIcons/pencil.svg';
import { ReactComponent as IconCheck } from '../../images/FontAwesomeIcons/check.svg';
import { ReactComponent as IconX } from '../../images/FontAwesomeIcons/x.svg';
import styles from './EditableField.module.scss';

export default class EditableField extends React.Component {
  static propTypes = {
    className: PropTypes.string,
    initialValue: PropTypes.string.isRequired,
    setValueCallback: PropTypes.func,
    tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  };

  static defaultProps = {
    className: '',
    setValueCallback: () => {},
    tag: 'span',
  };

  state = {
    isEditing: false,
    value: this.props.initialValue,
    valueCopy: this.props.initialValue,
  };

  setEditing = () => {
    const { value } = this.state;

    this.setState({ isEditing: true, valueCopy: value });
  };

  cancelEdit = () => {
    const { valueCopy } = this.state;

    this.setState({ isEditing: false, value: valueCopy });
  };

  saveEdit = () => {
    const { setValueCallback } = this.props;
    const { value } = this.state;

    this.setState({ isEditing: false, valueCopy: value }, () => setValueCallback(value));
  };

  onChange = event => {
    const { target } = event;
    this.setState({ value: target.value });
  };

  render() {
    const { className, tag: Tag } = this.props;
    const { isEditing, value } = this.state;

    return (
      <div className={classNames(styles.EditableField, className)}>
        {isEditing ? (
          <>
            <input value={value} onChange={this.onChange} className={styles.input} />

            <Button className={styles.button} onClick={this.saveEdit} type="submit">
              <ScreenReaderOnly>Save</ScreenReaderOnly>
              <IconCheck className={styles.icon} />
            </Button>

            <Button className={styles.button} onClick={this.cancelEdit} theme="secondary">
              <ScreenReaderOnly>Cancel</ScreenReaderOnly>
              <IconX className={styles.icon} />
            </Button>
          </>
        ) : (
          <>
            <Tag className={styles.value}>{value}</Tag>

            <Button className={styles.button} onClick={this.setEditing}>
              <ScreenReaderOnly>Edit</ScreenReaderOnly>
              <IconPencil className={styles.icon} />
            </Button>
          </>
        )}
      </div>
    );
  }
}
