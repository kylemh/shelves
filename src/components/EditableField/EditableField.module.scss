@import '../../styles/variables.scss';

.EditableField {
  color: $primary;
  display: flex;
  align-items: center;
  justify-content: flex-start;
}

.icon {
  height: 1rem;
}

.button {
  margin: 0 0 0 1rem;
  min-width: 2.5rem;
}

.input {
  font-size: 1rem;
  height: 2rem;
}
