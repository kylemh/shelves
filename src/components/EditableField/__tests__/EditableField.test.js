import React from 'react';
import { mount, shallow } from 'enzyme';
// import wait from '../../../utils/test-utils/wait';
import EditableField from '../EditableField';

describe('EditableField', () => {
  it('should render', () => {
    const wrapper = shallow(<EditableField initialValue="test" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should be renderable when it in edit state', () => {
    const wrapper = shallow(<EditableField initialValue="test" />);

    wrapper.setState({ isEditing: true });
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('input');

    expect(wrapper).toMatchSnapshot();
  });

  it('should open and close without a value change when clicking edit and then cancle', () => {
    const testValue = 'Trip-hop';
    const wrapper = shallow(<EditableField initialValue={testValue} />);

    wrapper.find('Button').simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('input');

    wrapper
      .find('Button')
      .last()
      .simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('Button');

    expect(
      wrapper
        .find('span')
        .first()
        .text()
    ).toStrictEqual(testValue);
  });

  it('should open and close without a value change when clicking edit and then submit', () => {
    const testValue = 'French House';
    const wrapper = shallow(<EditableField initialValue={testValue} />);

    wrapper.find('Button').simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('input');

    wrapper
      .find('Button')
      .first()
      .simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('Button');

    expect(
      wrapper
        .find('span')
        .first()
        .text()
    ).toStrictEqual(testValue);
  });

  it('should save value to final tag when submitting a changed input', () => {
    const wrapper = shallow(<EditableField initialValue="Alternative Rock" tag="h1" />);

    wrapper.find('Button').simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('input');

    const newValue = 'Classic Rock';
    wrapper.find('input').simulate('change', { target: { value: newValue } });
    wrapper
      .find('Button')
      .first()
      .simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('Button');

    expect(wrapper.find('h1').text()).toStrictEqual(newValue);
  });

  it('should fire passed callback submitting a changed input', () => {
    const mockedCallback = jest.fn();
    const wrapper = shallow(
      <EditableField initialValue="Pop" tag="h1" setValueCallback={mockedCallback} />
    );

    wrapper.find('Button').simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('input');

    wrapper.find('input').simulate('change', { target: { value: 'Disco' } });
    expect(mockedCallback).toBeCalledTimes(0);
    wrapper
      .find('Button')
      .first()
      .simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('Button');

    expect(mockedCallback).toBeCalledTimes(1);
  });

  it("should keep the initial value if a changed input's submission was cancelled", () => {
    const firstValue = 'Blues';
    const wrapper = mount(<EditableField initialValue={firstValue} tag="h1" />);

    wrapper.find('Button').simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('input');

    wrapper.find('input').simulate('change', { target: { value: 'Jazz' } });
    wrapper
      .find('Button')
      .last()
      .simulate('click');
    wrapper.update();
    expect(wrapper).toContainExactlyOneMatchingElement('Button');

    expect(wrapper.find('h1').text()).toStrictEqual(firstValue);
  });
});
