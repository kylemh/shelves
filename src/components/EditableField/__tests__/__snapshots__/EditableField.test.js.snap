// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableField should be renderable when it in edit state 1`] = `
<div
  className="EditableField"
>
  <input
    className="input"
    onChange={[Function]}
    value="test"
  />
  <Button
    className="button"
    disabled={false}
    fullWidth={false}
    onClick={[Function]}
    tabIndex={0}
    theme="primary"
    type="submit"
  >
    <ScreenReaderOnly>
      Save
    </ScreenReaderOnly>
    <ReactComponent
      className="icon"
    />
  </Button>
  <Button
    className="button"
    disabled={false}
    fullWidth={false}
    onClick={[Function]}
    tabIndex={0}
    theme="secondary"
    type="button"
  >
    <ScreenReaderOnly>
      Cancel
    </ScreenReaderOnly>
    <ReactComponent
      className="icon"
    />
  </Button>
</div>
`;

exports[`EditableField should render 1`] = `
<div
  className="EditableField"
>
  <span
    className="value"
  >
    test
  </span>
  <Button
    className="button"
    disabled={false}
    fullWidth={false}
    onClick={[Function]}
    tabIndex={0}
    theme="primary"
    type="button"
  >
    <ScreenReaderOnly>
      Edit
    </ScreenReaderOnly>
    <ReactComponent
      className="icon"
    />
  </Button>
</div>
`;
